File: HM/3_LongestSubstringWithoutRepeatingCharacters.py
def lengthOfLongestSubstring(s: str) -> int:
    slidingWindow = set()
    left = 0
    res = 0

    for right in range(len(s)):
        #오른쪽 포인터가 지금 슬라이딩 윈도우에 있을 경우 *왼포인터랑겹칠때
        while s[right] in slidingWindow:
            #왼 포인터를 슬라이딩 윈도우에서 삭제
            slidingWindow.remove(s[left])
            #왼 포인터를 한칸 옮김
            left += 1
        #그렇지 않으면, 슬라이딩 윈도우에 추가
        slidingWindow.add(s[right])
        #현재 슬라이딩 윈도우 길이가 더 길다면 res값을 업데이트
        res = max(res, len(slidingWindow))

    return res

File: HM/3_LongestSubstringWithoutRepeatingCharacters.c
#include <stdio.h>

int lengthOfLongestSubstring(char *s) {
    int charIndex[128] = {0};  // Assuming ASCII characters, initialize array to store the last index of each character
    int left = 0;
    int res = 0;

    for (int right = 0; s[right] != '\0'; right++) {
        // If the character at the right pointer is already in the sliding window,
        // move the left pointer to the right of the last occurrence of that character.
        if (charIndex[s[right]] > left) {
            left = charIndex[s[right]];
        }

        // Update the last index of the character in the array.
        charIndex[s[right]] = right + 1;

        // Update the length of the longest substring without repeating characters.
        if (right - left + 1 > res) {
            res = right - left + 1;
        }
    }

    return res;
}
